### PR TITLE
8.0 l10n ca account check writing

### PR DESCRIPTION
**Pull request by veloutin**
Attempt at porting l10n_ca_account_check_writing to v8.

res_currency_print_on_check is a dependency and needed
porting as well.

### CODE REVIEW

**Comment by max3903** [comment]
I got this error message when I tried to validate a payment:

ValueError: "'account.voucher' object has no attribute '_get_amount_in_word'" while evaluating
u'proforma_voucher()'

File: l10n_ca_account_check_writing/account_voucher.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+            amount_in_word = self._get_amount_in_word(cr, uid, i_id,
+                                                      context=context)
+            self.write(cr, uid, i_id, {'amount_in_word': amount_in_word},
+                       context=context)
+
+        return super(AccountVoucher, self).proforma_voucher(cr, uid, ids,
+                                                            context=context)
+
+
+class VoucherLine(models.Model):
```
Comment (bwrsandman) [inline l10n_ca_account_check_writing/account_voucher.py]: This class should go into its own file

File: l10n_ca_account_check_writing/company.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+class Company(models.Model):
+    _inherit = _name = 'res.company'
+
+    check_layout = fields.Selection([
+        ('top', 'Check on Top'),
+        ('middle', 'Check in middle'),
+        ('bottom', 'Check on bottom'),
+        ('top_ca', 'Check on top (CA)'),
+        ('middle_ca', 'Check in middle (CA)'),
+        #  ('bottom_ca', 'Check on bottom (CA)'),
```
Comment (bwrsandman) [inline l10n_ca_account_check_writing/company.py]: Please remove this line

File: l10n_ca_account_check_writing/report/l10n_ca_check_print.py [hunk 3]
```diff
@@ -56,10 +58,14 @@ def get_lines(self, voucher_lines):
                     res = {
                         'date_due': voucher_line.date_due,
                         'name': name,
-                        'amount_original': voucher_line.amount_original and voucher_line.amount_original or False,
-                        'amount_unreconciled': voucher_line.amount_unreconciled and voucher_line.amount_unreconciled
-                        or False,
-                        'amount': voucher_line.amount and voucher_line.amount or False,
+                        'amount_original': (
+                            voucher_line.amount_original and
+                            voucher_line.amount_original or False),
```
Comment (bwrsandman) [inline l10n_ca_account_check_writing/report/l10n_ca_check_print.py]: `voucher_line.amount_original and voucher_line.amount_original or False` can be shortened to `voucher_line.amount_original or False`

File: l10n_ca_account_check_writing/report/l10n_ca_check_print.py [hunk 4]
```diff
@@ -56,10 +58,14 @@ def get_lines(self, voucher_lines):
...
+                        'amount_unreconciled': (
+                            voucher_line.amount_unreconciled and
+                            voucher_line.amount_unreconciled or False),
```
Comment (bwrsandman) [inline l10n_ca_account_check_writing/report/l10n_ca_check_print.py]: `voucher_line.amount_unreconciled and voucher_line.amount_unreconciled or False` can be shortened to `voucher_line.amount_unreconciled or False`

File: l10n_ca_account_check_writing/report/l10n_ca_check_print.py [hunk 5]
```diff
@@ -56,10 +58,14 @@ def get_lines(self, voucher_lines):
...
+                        'amount': (voucher_line.amount and
+                                   voucher_line.amount or False),
```
Comment (bwrsandman) [inline l10n_ca_account_check_writing/report/l10n_ca_check_print.py]: `voucher_line.amount and voucher_line.amount or False` can be shortened to `voucher_line.amount or False`

File: l10n_ca_account_check_writing/report/l10n_ca_check_print.py [hunk 6]
```diff
@@ -85,7 +91,7 @@ def get_lines(self, voucher_lines):
 report_sxw.report_sxw(
     'report.l10n.ca.account.print.check.middle',
     'account.voucher',
-    'addons/l10n_ca_account_check_writing/report/l10n_ca_check_print_middle.rml',
+    'addons/l10n_ca_account_check_writing/report/l10n_ca_check_print_middle.rml',  # NOQA
```
Comment (bwrsandman) [inline l10n_ca_account_check_writing/report/l10n_ca_check_print.py]: Please add a reason why noqa is used here, if it's a too long line, you can just break it in half

```
    'addons/l10n_ca_account_check_writing/'
    'report/l10n_ca_check_print_middle.rml'
```

File: l10n_ca_account_check_writing/company.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+from openerp import fields, models
+
+
+class Company(models.Model):
+    _inherit = _name = 'res.company'
```
Comment (bwrsandman) [inline l10n_ca_account_check_writing/company.py]: `_name =` should not be necessary

File: l10n_ca_account_check_writing/account_voucher_line.py [hunk 8]
```diff
@@ -20,18 +20,22 @@
...
-        ], "Choose Check layout")
-    }
+class VoucherLine(models.Model):
+    _inherit = _name = 'account.voucher.line'
+
+    @api.multi
+    def get_suppl_inv_num(self):
+        move_obj = self.env['account.move.line']
+        for rec in self:
+            move_line = move_obj.browse(rec.move_line_id.id)
```
Comment (bwrsandman) [inline l10n_ca_account_check_writing/account_voucher_line.py]: Seems a bit redundant, why not just `move_line = rec.move_line_id`

File: l10n_ca_account_check_writing/account_voucher_line.py [hunk 9]
```diff
@@ -20,18 +20,22 @@
...
-            ('top', 'Check on Top'),
-            ('middle', 'Check in middle'),
-            ('bottom', 'Check on bottom'),
-            ('top_ca', 'Check on top (CA)'),
-            ('middle_ca', 'Check in middle (CA)'),
-            #  ('bottom_ca', 'Check on bottom (CA)'),
-        ], "Choose Check layout")
-    }
+class VoucherLine(models.Model):
+    _inherit = _name = 'account.voucher.line'
```
Comment (bwrsandman) [inline l10n_ca_account_check_writing/account_voucher_line.py]: `_name =` should not be necessary

File: l10n_ca_account_check_writing/account_voucher.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+            'datas': {
+                'model': 'account.voucher',
+                'id': ids and ids[0] or False,
+                'ids': ids and ids or [],
+                'report_type': 'pdf'
+            },
+            'nodestroy': True
+        }
+
+    def proforma_voucher(self, cr, uid, ids, context=None):
```
Comment (bwrsandman) [inline l10n_ca_account_check_writing/account_voucher.py]: Do you think you could factor this into the new API?

File: l10n_ca_account_check_writing/account_voucher.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+
+
+def custom_translation(s, lang):
+    # Odoo uses the current stack frame, yes, the *stack frame* to determine
+    # which language _() should translate a string in. If we want to translate
+    # a string in another language, such as a supplier's language, we have to
+    # resort to hacks such as this one. "context" is sought after in the
+    # stackframe, so we have to set it.
+    context = {'lang': lang}  # NOQA
+    return _(s)
```
Comment (bwrsandman) [inline l10n_ca_account_check_writing/account_voucher.py]: This function is unnecessary with v8 environment switching

File: l10n_ca_account_check_writing/account_voucher.py [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+            amount_in_word = num2words(int(amount))
+
+        currency_name = currency.print_on_check
+        cents = int(amount * 100) % 100
+        total_length = len(amount_in_word) + len(currency_name)
+        if total_length < 67:
+            stars = '*' * (67 - total_length)
+        else:
+            stars = ''
+        AND = custom_translation("and", lang)
```
Comment (bwrsandman) [inline l10n_ca_account_check_writing/account_voucher.py]: You should be able to get custom translations using v8 environment context switching

**Comment by max3903** [comment]
I got this error message when printing a check. I tried on top and in the middle, same error.

2014-11-25 03:07:01,185 6603 ERROR Meetup openerp.service.report: Exception: 'ascii' codec can't encode character u'\xe9' in position 1: ordinal not in range(128)
Traceback (most recent call last):
  File "/home/max/odoo/meetup/git/odoo/openerp/service/report.py", line 93, in go
    result, format = openerp.report.render_report(cr, uid, ids, object, datas, context)
  File "/home/max/odoo/meetup/git/odoo/openerp/report/init.py", line 40, in render_report
    return registry['ir.actions.report.xml'].render_report(cr, uid, ids, name, data, context)
  File "/home/max/odoo/meetup/git/odoo/openerp/api.py", line 237, in wrapper
    return old_api(self, _args, *_kwargs)
  File "/home/max/odoo/meetup/git/odoo/openerp/addons/base/ir/ir_actions.py", line 155, in render_report
    return new_report.create(cr, uid, res_ids, data, context)
  File "/home/max/odoo/meetup/git/odoo/openerp/report/report_sxw.py", line 392, in create
    fnct_ret = fnct(cr, uid, ids, data, report_xml, context)
  File "/home/max/odoo/meetup/git/odoo/openerp/report/report_sxw.py", line 461, in create_source_pdf
    return self.create_single_pdf(cr, uid, ids, data, report_xml, context)
  File "/home/max/odoo/meetup/git/odoo/openerp/report/report_sxw.py", line 483, in create_single_pdf
    pdf = create_doc(etree.tostring(processed_rml),rml_parser.localcontext,logo,title.encode('utf8'))
  File "/home/max/odoo/meetup/git/odoo/openerp/report/interface.py", line 221, in create_pdf
    obj.render()
  File "/home/max/odoo/meetup/git/odoo/openerp/report/render/render.py", line 59, in render
    self._result = self._render()
  File "/home/max/odoo/meetup/git/odoo/openerp/report/render/rml.py", line 41, in _render
    return rml2pdf.parseNode(self.rml, self.localcontext, images=self.bin_datas, path=self.path,title=self.title)
  File "/home/max/odoo/meetup/git/odoo/openerp/report/render/rml2pdf/trml2pdf.py", line 1046, in parseNode
    r.render(fp)
  File "/home/max/odoo/meetup/git/odoo/openerp/report/render/rml2pdf/trml2pdf.py", line 335, in render
    pt_obj.render(el)
  File "/home/max/odoo/meetup/git/odoo/openerp/report/render/rml2pdf/trml2pdf.py", line 1026, in render
    self.doc_tmpl.build(fis)
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/doctemplate.py", line 890, in build
    self.handle_flowable(flowables)
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/doctemplate.py", line 773, in handle_flowable
    if frame.add(f, canv, trySplit=self.allowSplitting):
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/frames.py", line 190, in _add
    flowable.drawOn(canv, self._x + self._leftExtraIndent, y, _sW=aW-w)
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/flowables.py", line 110, in drawOn
    self._drawOn(canvas)
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/flowables.py", line 91, in _drawOn
    self.draw()#this is the bit you overload
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/tables.py", line 1368, in draw
    self._drawCell(cellval, cellstyle, (colpos, rowpos), (colwidth, rowheight))
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/tables.py", line 1459, in _drawCell
    w, h = self._listCellGeom(cellval,colwidth,cellstyle,W=W, H=H,aH=rowheight)
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/tables.py", line 397, in _listCellGeom
    vw, vh = v.wrapOn(canv, aW, aH)
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/flowables.py", line 121, in wrapOn
    w, h = self.wrap(aW,aH)
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/tables.py", line 1173, in wrap
    self._calc(availWidth, availHeight)
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/tables.py", line 625, in _calc
    self._calc_height(availHeight,availWidth,W=W)
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/tables.py", line 556, in _calc_height
    dW,t = self._listCellGeom(v,w or self._listValueWidth(v),s)
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/tables.py", line 397, in _listCellGeom
    vw, vh = v.wrapOn(canv, aW, aH)
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/flowables.py", line 121, in wrapOn
    w, h = self.wrap(aW,aH)
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/paragraph.py", line 1036, in wrap
    blPara = self.breakLines([first_line_width, later_widths])
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/paragraph.py", line 1244, in breakLines
    words[0:0] = _splitWord(word,maxWidth-spaceWidth-currentWidth,maxWidths,lineno,fontName,fontSize,self.encoding)
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/paragraph.py", line 553, in _splitWord
    R.append(_SplitText(wordText))
UnicodeEncodeError: 'ascii' codec can't encode character u'\xe9' in position 1: ordinal not in range(128)
2014-11-25 03:07:01,273 6603 ERROR Meetup openerp.addons.web.controllers.main: An exception occured during an http request
Traceback (most recent call last):
  File "/home/max/odoo/meetup/git/odoo/addons/web/controllers/main.py", line 69, in wrap
    return f(_args, *_kwargs)
  File "/home/max/odoo/meetup/git/odoo/addons/web/controllers/main.py", line 1576, in index
    request.session.db, request.session.uid, request.session.password, report_id)
  File "/home/max/odoo/meetup/git/odoo/openerp/http.py", line 816, in proxy_method
    result = dispatch_rpc(self.service_name, method, args)
  File "/home/max/odoo/meetup/git/odoo/openerp/http.py", line 100, in dispatch_rpc
    result = dispatch(method, params)
  File "/home/max/odoo/meetup/git/odoo/openerp/service/report.py", line 34, in dispatch
    res = fn(db, uid, _params)
  File "/home/max/odoo/meetup/git/odoo/openerp/service/report.py", line 141, in exp_report_get
    return _check_report(report_id)
  File "/home/max/odoo/meetup/git/odoo/openerp/service/report.py", line 119, in _check_report
    raise openerp.osv.orm.except_orm(exc.message, exc.traceback)
except_orm: (u'ascii\nR\xe9ducti\n1\n2\nordinal not in range(128)', (<type 'exceptions.UnicodeEncodeError'>, UnicodeEncodeError('ascii', u'R\xe9ducti', 1, 2, 'ordinal not in range(128)'), <traceback object at 0x7f5161a2a200>))2014-11-25 03:07:01,185 6603 ERROR Meetup openerp.service.report: Exception: 'ascii' codec can't encode character u'\xe9' in position 1: ordinal not in range(128)
Traceback (most recent call last):
  File "/home/max/odoo/meetup/git/odoo/openerp/service/report.py", line 93, in go
    result, format = openerp.report.render_report(cr, uid, ids, object, datas, context)
  File "/home/max/odoo/meetup/git/odoo/openerp/report/init.py", line 40, in render_report
    return registry['ir.actions.report.xml'].render_report(cr, uid, ids, name, data, context)
  File "/home/max/odoo/meetup/git/odoo/openerp/api.py", line 237, in wrapper
    return old_api(self, *args, *_kwargs)
  File "/home/max/odoo/meetup/git/odoo/openerp/addons/base/ir/ir_actions.py", line 155, in render_report
    return new_report.create(cr, uid, res_ids, data, context)
  File "/home/max/odoo/meetup/git/odoo/openerp/report/report_sxw.py", line 392, in create
    fnct_ret = fnct(cr, uid, ids, data, report_xml, context)
  File "/home/max/odoo/meetup/git/odoo/openerp/report/report_sxw.py", line 461, in create_source_pdf
    return self.create_single_pdf(cr, uid, ids, data, report_xml, context)
  File "/home/max/odoo/meetup/git/odoo/openerp/report/report_sxw.py", line 483, in create_single_pdf
    pdf = create_doc(etree.tostring(processed_rml),rml_parser.localcontext,logo,title.encode('utf8'))
  File "/home/max/odoo/meetup/git/odoo/openerp/report/interface.py", line 221, in create_pdf
    obj.render()
  File "/home/max/odoo/meetup/git/odoo/openerp/report/render/render.py", line 59, in render
    self._result = self._render()
  File "/home/max/odoo/meetup/git/odoo/openerp/report/render/rml.py", line 41, in _render
    return rml2pdf.parseNode(self.rml, self.localcontext, images=self.bin_datas, path=self.path,title=self.title)
  File "/home/max/odoo/meetup/git/odoo/openerp/report/render/rml2pdf/trml2pdf.py", line 1046, in parseNode
    r.render(fp)
  File "/home/max/odoo/meetup/git/odoo/openerp/report/render/rml2pdf/trml2pdf.py", line 335, in render
    pt_obj.render(el)
  File "/home/max/odoo/meetup/git/odoo/openerp/report/render/rml2pdf/trml2pdf.py", line 1026, in render
    self.doc_tmpl.build(fis)
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/doctemplate.py", line 890, in build
    self.handle_flowable(flowables)
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/doctemplate.py", line 773, in handle_flowable
    if frame.add(f, canv, trySplit=self.allowSplitting):
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/frames.py", line 190, in _add
    flowable.drawOn(canv, self._x + self._leftExtraIndent, y, _sW=aW-w)
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/flowables.py", line 110, in drawOn
    self._drawOn(canvas)
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/flowables.py", line 91, in _drawOn
    self.draw()#this is the bit you overload
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/tables.py", line 1368, in draw
    self._drawCell(cellval, cellstyle, (colpos, rowpos), (colwidth, rowheight))
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/tables.py", line 1459, in _drawCell
    w, h = self._listCellGeom(cellval,colwidth,cellstyle,W=W, H=H,aH=rowheight)
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/tables.py", line 397, in _listCellGeom
    vw, vh = v.wrapOn(canv, aW, aH)
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/flowables.py", line 121, in wrapOn
    w, h = self.wrap(aW,aH)
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/tables.py", line 1173, in wrap
    self._calc(availWidth, availHeight)
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/tables.py", line 625, in _calc
    self._calc_height(availHeight,availWidth,W=W)
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/tables.py", line 556, in _calc_height
    dW,t = self._listCellGeom(v,w or self._listValueWidth(v),s)
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/tables.py", line 397, in _listCellGeom
    vw, vh = v.wrapOn(canv, aW, aH)
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/flowables.py", line 121, in wrapOn
    w, h = self.wrap(aW,aH)
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/paragraph.py", line 1036, in wrap
    blPara = self.breakLines([first_line_width, later_widths])
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/paragraph.py", line 1244, in breakLines
    words[0:0] = _splitWord(word,maxWidth-spaceWidth-currentWidth,maxWidths,lineno,fontName,fontSize,self.encoding)
  File "/usr/lib/python2.7/dist-packages/reportlab/platypus/paragraph.py", line 553, in _splitWord
    R.append(_SplitText(wordText))
UnicodeEncodeError: 'ascii' codec can't encode character u'\xe9' in position 1: ordinal not in range(128)
2014-11-25 03:07:01,273 6603 ERROR Meetup openerp.addons.web.controllers.main: An exception occured during an http request
Traceback (most recent call last):
  File "/home/max/odoo/meetup/git/odoo/addons/web/controllers/main.py", line 69, in wrap
    return f(_args, *_kwargs)
  File "/home/max/odoo/meetup/git/odoo/addons/web/controllers/main.py", line 1576, in index
    request.session.db, request.session.uid, request.session.password, report_id)
  File "/home/max/odoo/meetup/git/odoo/openerp/http.py", line 816, in proxy_method
    result = dispatch_rpc(self.service_name, method, args)
  File "/home/max/odoo/meetup/git/odoo/openerp/http.py", line 100, in dispatch_rpc
    result = dispatch(method, params)
  File "/home/max/odoo/meetup/git/odoo/openerp/service/report.py", line 34, in dispatch
    res = fn(db, uid, *params)
  File "/home/max/odoo/meetup/git/odoo/openerp/service/report.py", line 141, in exp_report_get
    return _check_report(report_id)
  File "/home/max/odoo/meetup/git/odoo/openerp/service/report.py", line 119, in _check_report
    raise openerp.osv.orm.except_orm(exc.message, exc.traceback)
except_orm: (u'ascii\nR\xe9ducti\n1\n2\nordinal not in range(128)', (<type 'exceptions.UnicodeEncodeError'>, UnicodeEncodeError('ascii', u'R\xe9ducti', 1, 2, 'ordinal not in range(128)'), <traceback object at 0x7f5161a2a200>))

**Comment by max3903** [comment]
Results of my tests:
- Top: https://dl.dropboxusercontent.com/u/24473318/openerp/check_on_top.pdf
- Middle: https://dl.dropboxusercontent.com/u/24473318/openerp/check_in_middle.pdf

Common to the 2 reports
- [x] Supplier refunds amount are positive. They should be negative.
- [x] The currency names are missing in the translation file of res_currency_print_on_check

Check on top:
- [x] 'DATE' is on 2 lines
- [x] Useless 'YYYY-MM-DD' should be removed
- [x] 'MEMO' is on 2 lines

Check in the middle:
- [x] the date is on 2 lines in the top table and in the check
- [x] Amount in digits are not aligned with amount in letters.

Recheck the result against the standard:
http://www.cdnpay.ca/imis15/pdf/pdfs_rules/standard_006_fr.pdf (page 31)

**Comment by max3903** [comment]
- [x] Use the default template set on the company in the form views
- On the company, change the default template for check
- Go to "Supplier payments" or "Write check" menu items
- Open a record, click on "Print" and  then "Check"
- You don't get the default template

**Comment by max3903** [comment]
Page 22 du standard :
- Les astérisques, s’ils sont utilisés dans la zone du montant en chiffres, ne
  doivent être imprimés qu’avant le montant en chiffres (p. ex., $****_45.00 ou
  ***_*45,00$). L’utilisation des astérisques dans toute autre position dans la zone
  du montant en chiffres n’est pas autorisée.
- Les astérisques, s’ils sont utilisés dans la zone du montant en lettres, ne doivent
  être imprimés qu’à la gauche du montant en lettres (p. ex., *********quarante-cinq
  dollars). L’utilisation d’astérisques dans toute autre position dans la zone du
  montant en lettres n’est pas autorisée.

**Comment by max3903** [comment]
:+1:

**Comment by JehoG** [comment]
What's the status of this PR?

**Comment by max3903** [comment]
The amount in letter is not in the report if the payment is not posted. It should be whatever the payment state.

**Comment by foutoucour** [comment]
:+1:

**Pull request merged**